### DISCUSSION TITLE
Neomutt completion API

### DISCUSSION THREAD

**Discussion by Amudtogal** [neomutt/neomutt#3955] [Development]
Following up from a discussion with @flatcap and some experimental development on my side, I am proposing the new completion API for neomutt.

The plan is to replace all the different spots in neomutt, where users can complete filenames/aliases/notmuch tags etc., with a unified completion API.

The current state of the completion can be found at [here](https://github.com/neomutt/completion-api), and features a bunch of tests and different completion modes (fuzzy, regex, exact), as well as a state machine that allows to "tab" through different matches one at a time.
It is not yet integrated into neomutt itself (I am working on that), but before I open a pull request, I would like to get some feedback and discuss a few key questions.

But let's lead with an example:

```
  // we need to set the locale settings, otherwise UTF8 chars won't work as expected
  setlocale(LC_ALL, "en_US.UTF-8");
  Completion *comp = compl_new(COMPL_MODE_EXACT);

  // adding possible matches to the completion
  compl_add(comp, "apfel", 6);
  compl_add(comp, "apple", 6);
  compl_add(comp, "apply", 6);
  compl_add(comp, "arange", 7);

  // user types "ap"
  compl_type(comp, "ap", 3);

  // with the compl_complete, we request a completion (could be bound to <tab>)
  char *result = compl_complete(comp);
  TEST_CHECK(STR_EQ(result, "apfel"));

  // completing again will cycle through the match results
  result = compl_complete(comp);
  TEST_CHECK(STR_EQ(result, "apple"));

  result = compl_complete(comp);
  TEST_CHECK(STR_EQ(result, "apply"));

  // in COMPL_MODE_EXACT, arange is not a match, so we reset to the user typed string
  result = compl_complete(comp);
  TEST_CHECK(STR_EQ(result, "ap"));

  // once we are done, we clean up the completion
  compl_free(comp);
```

The completion has currently three matching modes:
- `COMPL_MODE_EXACT`: characters need to match exactly
- `COMPL_MODE_FUZZY`: match "distance" is computed by counting insertion/deletions/substitution/transposition of single characters (damereau-levenshtein distance, see [here](https://dl.acm.org/doi/10.1145/375360.375365) for an introduction)
- `COMPL_MODE_REGEX`: we use `regex.h` (regcomp) to compute regular expression matching

For all of these modes, each entry in the completion list gets assigned a match distance (int), which is lower for closer matches.
For the exact matching, this is the count of characters needed to be added to the typed string.
For regex this is currently a naive implementation (length difference between regular expression and match).
For fuzzy matching, this resolves to the damerau-levenshtein distance.

All of the entries are then sorted according to their match distance (closest first), and then alphabetically, leaving no-matches at the end of the list.
The `COMPL_MODE` can be further configured by the `COMPL_MATCH` flags (e.g. `COMPL_MATCH_IGNORECASE`).

---

For the next steps, I would like to get some input on these questions:

- for the different use cases, we need to be able to gather all the potential matches (e.g. all the filenames in a given directory), to gather them in an array, or add them one by one to the completion: any ideas on where best to put these?
- for users, it would be great to have some initial completion defaults, but changing them on the go would be great.

@flatcap has gathered all the potential touch points of the completion API (calls to `mw_get_field`):
```
alias/alias.c:498:        mw_get_field(_("Save to file: "), buf, MUTT_COMP_FILE | MUTT_COMP_CLEAR,
attach/recvattach.c:303:  mw_get_field(prompt, buf, MUTT_COMP_FILE | MUTT_COMP_CLEAR, false,
attach/recvattach.c:484:  mw_get_field(_("Save to file: "), buf, MUTT_COMP_FILE | MUTT_COMP_CLEAR,
attach/recvattach.c:732:  mw_get_field((filter ? _("Filter through: ") : _("Pipe to: ")),
browser/functions.c:99:   mw_get_field(_("New file name: "), buf, MUTT_COMP_FILE, false,
browser/functions.c:313:  mw_get_field(_("Chdir to: "), buf, MUTT_COMP_FILE, false, NULL, NULL, NULL);
browser/functions.c:940:  mw_get_field(tmp2, buf, MUTT_COMP_PATTERN, false, NULL, NULL, NULL) != 0) ||
compose/functions.c:1310: mw_get_field(_("New file: "), fname, MUTT_COMP_FILE, false, NULL, NULL, NULL) != 0) ||
compose/functions.c:1420: mw_get_field(_("Send attachment with name: "), fname, MUTT_COMP_FILE,
compose/functions.c:1752: mw_get_field(_("Rename to: "), fname, MUTT_COMP_FILE, false, NULL, NULL, NULL) == 0) &&
envelope/functions.c:120: mw_get_field(_(Prompts[field]), new_list, MUTT_COMP_ALIAS, false, NULL, NULL, NULL) == 0)
envelope/functions.c:230: mw_get_field(Prompts[HDR_FCC], fname, MUTT_COMP_FILE | MUTT_COMP_CLEAR,
external.c:121:           mw_get_field(buf_string(prompt), buf, MUTT_COMP_ALIAS, false, NULL, NULL, NULL);
external.c:431:           mw_get_field(_("Pipe to command: "), buf, MUTT_COMP_FILE_SIMPLE, false,
external.c:601:           mw_get_field(_("Shell command: "), buf, MUTT_COMP_FILE_SIMPLE, false,
external.c:645:           mw_get_field(":", buf, MUTT_COMP_COMMAND, false, NULL, NULL, NULL) != 0) ||
gui/curs_lib.c:513:       mw_get_field(pc, fname, (mailbox ? MUTT_COMP_FILE_MBOX : MUTT_COMP_FILE) | MUTT_COMP_CLEAR,
imap/browse.c:410:        mw_get_field(_("Create mailbox: "), name, MUTT_COMP_FILE, false, NULL, NULL, NULL) != 0)
imap/browse.c:469:        mw_get_field(buf_string(buf), newname, MUTT_COMP_FILE, false, NULL, NULL, NULL) != 0)
index/functions.c:2716:   mw_get_field("Query: ", buf, MUTT_COMP_NM_QUERY, false, NULL, NULL, NULL) != 0) ||
mutt_header.c:145:        mw_get_field("Label: ", buf, MUTT_COMP_LABEL, false, NULL, NULL, NULL) != 0)
muttlib.c:624:            mw_get_field(_("File under directory: "), tmp, MUTT_COMP_FILE | MUTT_COMP_CLEAR,
notmuch/notmuch.c:2396:   mw_get_field("Add/remove labels: ", buf, MUTT_COMP_NM_TAG, false, NULL, NULL, NULL) != 0)
pager/functions.c:315:    mw_get_field(((op == OP_SEARCH) || (op == OP_SEARCH_NEXT)) ? _("Search for: ")
pattern/pattern.c:202:    mw_get_field(prompt, buf, MUTT_COMP_PATTERN | MUTT_COMP_CLEAR, false,
pattern/pattern.c:308:    mw_get_field(prompt, buf, MUTT_COMP_PATTERN | MUTT_COMP_CLEAR, false,
pattern/pattern.c:458:    mw_get_field(((op == OP_SEARCH) || (op == OP_SEARCH_NEXT)) ? _("Search for: ")
pattern/pattern.c:624:    mw_get_field(((op == OP_SEARCH) || (op == OP_SEARCH_NEXT)) ? _("Search for: ")
recvcmd.c:209:            mw_get_field(buf_string(prompt), buf, MUTT_COMP_ALIAS, false, NULL, NULL, NULL) != 0) ||
send/send.c:190:          mw_get_field(field, buf, MUTT_COMP_ALIAS, false, NULL, NULL, NULL) != 0)
```
If you know/think of any other potential users of the completion, let me know.

I hope I have explained everything well enought, and I look forward to some feedback!

Cheers,
Simon

**Comment by lucilanga** [comment]
Great topic!

> * for the different use cases, we need to be able to gather all the potential matches (e.g. all the filenames in a given directory), to gather them in an array, or add them one by one to the completion: any ideas on where best to put these?

Are you asking this because it's a potentially costly operation that you need to do it anyway ?
And by asking where to put it you actually mean when to do it ?
eg. before mw_get_field, inside mw_get_field, after <tab>  ?

As for array vs multiple compl_add. Does it really matter ? I mean you already do ARRAY_FOREACH with compl_add() on compl_from_array. Is this going to be different ?



> * for users, it would be great to have some initial completion defaults, but changing them on the go would be great.

like going from icase to case runtime ? That would be nice, but how do you actually see this ? binding another key ? like ctrl tab or so ? or perhaps prefix the search with something ?

**Reply by Amudtogal** [reply]
> Are you asking this because it's a potentially costly operation that you need to do it anyway ?

Well, not really. For a completion it always needs to be done anyway.

> And by asking where to put it you actually mean when to do it ?
> eg. before mw_get_field, inside mw_get_field, after ?

Yes. We do need different "collection" functions that gather the items that could be valid completions, e.g. gather_filenames/gather_aliases/gather_email_addresses, and my question is whether we want to place these in the completion library as well.
Benefit: easy to find, as all completion-stuff is in one place.
Disadvantage: other parts of neomutt might rely on these functions as well (e.g. to gather the list of filenames for the file selector)...
So it is rather a question of style and consistency...

> As for array vs multiple compl_add. Does it really matter ?

No, you are right, I would probably only keep the `compl_add` anyway... (Just offering the array initialiser in case someone can come up with a smart usecase)

> like going from icase to case runtime ?

Yes, this is possible.

> That would be nice, but how do you actually see this ? binding another key ? like ctrl tab or so ? or perhaps prefix the search with something ?

I was thinking of a search window in the long run, which has some bindings that could be assigned by users, e.g. a <ctl-c> to toggle case sensitivity.
Although we might need to see how well this performs, as changing the flags requires a recomputation of all the match distances and sorting of the match list (for smaller completions not that big of an issue, but I do not know yet how well it scales).

**Comment by Amudtogal** [comment]
While further going through the code, I realized that we should actually initialize the `Completion` along with the editor.
At that point we already know what kind of data the user is entering (right?), and can fill the completion with items right away.
Then, on each <tab>, we will only update the typed string and complete it according to the flags.

I will have a look where I need to attach the `Completion` to...

**Reply by Amudtogal** [reply]
Probably best to place it as part of `EnterWindowData`, along with the completion flags.

**Comment by Amudtogal** [comment]
Another thing: can we watch the user typed string for changes?
We only need to recompute the match list after *new* user input, is there anything like that already in place?
Like an "edit" flag that is set whenever user edits the entered text...

**Comment by flatcap** [comment]
Read with a pinch of salt :-)

---

> I realized that we should actually initialize the Completion along with the editor

Yes.
Currently, nothing exists until the user hits <kbd>\<tab\></kbd>.
In the future, we want the caller to set up completion and pass that to `mw_get_field()`.

> At that point we already know what kind of data the user is entering (right?),
> and can fill the completion with items right away

Yes, the caller knows exactly what's going on.
And yes, we can generate **simple** completion entries ahead of time.

> Then, on each, we will only update the typed string and complete it according to the flags

I think it's important to make a distinction here.

I don't think we want the Editor to complete from a list.
I think we want the Completion code to manipulate the Editor.
The Editor wants to be ignorant of how completion works.

The Editor will keep a data pointer for the Completion and will make its data available for manipulation.

This way, Completion is free to do anything:
- Do nothing
- Complete a string
- Start a selection Dialog
- Display a popup menu

> I will have a look where I need to attach the Completion to..
> Probably best to place it as part of EnterWindowData, along with the completion flags

Yes, I think a single opaque struct pointer.

> Another thing: can we watch the user typed string for changes?

Currently no.

> We only need to recompute the match list after new user input,

Do we need to?
If the user's typing, then I think we should wait until the next `<complete>`.

Unless, Completion has taken over, e.g. a popup.

> is there anything like that already in place?
> Like an "edit" flag that is set whenever user edits the entered text..

We could certainly add more functions to the Completion API, so that it's aware of everything that the user is doing.

**Reply by Amudtogal** [reply]
Alright, I was on my way of incorporating the completion into `mw_get_field`, but actually setting it up outside makes much more sense.

The only issue I could see is that a major change in the user's text, could lead to a change in the way the data is added to the completion:

e.g. user is picking a file from a file tree (dir `/abc/` contains 40 or files/subfolders, dir `/abc/def` another 40 or so):
- starting from directory `/abc/de` they `<tab>` for completion
- they select the subdirectory `/abc/def/` and hit `<tab>` again
- at this point, they could be presented with a selection menu of files in that folder and complete their selection here.

As the completion list is built before we open `mw_get_field`, we would have to parse the whole tree of files to succeed in this task.
That would mean we get a huge list of files (depending on scenario), and we need to compute the match distance for each of those entries for every tab.
A smarter approach here (in my opinion), would be rebuilding the list of selections, once we have entered the first subdirectory.
But this would require to call the `gather_files_for_completion` (or however we would name that) function, while `mw_get_field` is still active.

I know that *premature optimisation is the root of all evil*, but I would like to be considerate with the amount of computations we force neomutt through :smile:

**Reply by Amudtogal** [reply]
My approach would be:

- initialise completion withing `mw_get_field`, using the existing flags (e.g. `MUTT_COMP_ALIAS`), which trigger the respective "gathering" function
- completion and editor interact back and forth (completion would mostly update the currently entered string on a `<tab>`)
- completion might rebuild its list of available completions in specific circumstances (eg. entering a subdirectory)
- completion is closed when editor exits

This would be a more shallow approach to completion: we only complete the first level of hierarchical structures (like filetrees).

However, this could mean that searching for `**/*.c` files with regex would circumvent the shallow approach (as we need to see all files here).

It depends on what we want though:
A) `file.txt` (for a list of `/abc/file.txt`, `/def/file.txt`, `def/ghi/file.txt`) -> `<tab>` -> no match (as we would have to pick the first layer first: `abc`, `def`)
B) `file.txt` (same list as above) -> `<tab>` -> `/abc/file.txt` -> `<tab>`  -> `/def/file.txt` -> `<tab>` -> `/def/ghi/file.txt`

Both are valid ways for a completion to work, but with a different approach "under the hood"

**Reply by Amudtogal** [reply]
So what shall it be: `shallow` completion, or `deep` completion? :smile:

**Comment by flatcap** [comment]
Here's the layout of functions in Auto-Completion:

<img src="https://flatcap.org/mutt/completion2.svg">

**Source**: [graphviz](https://flatcap.org/mutt/completion2.gv), [svg](https://flatcap.org/mutt/completion2.svg)

**Comment by flatcap** [comment]
OK, time for a detailed walk-through...

Let's start with `get_source_file()` and a general "file completion" service.

```c
get_source_file(buf, start_dir)
{
  struct Completion *comp = comp_file_new(CF_FILES, "*.c", start_dir);
  mw_get_field(prompt, comp, buf);
  comp_free(&comp);
}
```

`struct Completion` will be shared across all completion services.
It might look like:
```c
struct Completion
{
  int type;                      ///< Completion type
  struct CompletionAPI *api;     ///< Functions this API supports
  struct StringArray matches;    ///< Current results
  int selection;                 ///< Currently selected item
  char *word;                    ///< Current word being completed
  int num_tabs;                  ///< Number of times <tab>'s been hit
  void *cdata;                   ///< Private completion data
};
```

The Completion API might look like:
```c
struct CompletionAPI
{
  int tab_pressed(struct Completion *comp);  ///< Perform work
  void free_cdata(void **ptr);               ///< Function to free private data
};
```

We'll need more functions later, but that may be enough to start.

Now, when the user hits <kbd>\<tab\></kbd>, `op_editor_complete()` is called.
It immediately delegates:

```c
int op_editor_complete(struct EnterWindowData *wdata, int op)
{
  comp_complete(wdata->comp, wdata->buf, wdata->pos);
  return FR_SUCCESS;
}
```

`comp_complete()` can keep track of the the state of the buffer.
The first time it's called, it will delegate to the completion service.

`comp_file_tab_pressed()` will use its setup data to generate an initial set of matches.
It populates the `comp.matches` array.

When the user hits <kbd>\<tab\></kbd> again, we have a choice.

For the simple cases -- a fixed list of matches -- the core completion code can do all the work.
It picks the next entry and wraps around if necessary.

For dynamic cases, e.g. a directory is selected, it will need to update the "current word"
and regenerate the array of matches.

**Comment by flatcap** [comment]
However, first we need to move the completion functions: `complete_*()`

They all have the same prototype, so `op_editor_complete()` shouldn't care which is used.

**Steps**:
- Export the functions:
    - De-`static` them
    - Declare in header
    - Create a `typedef` for the prototype
 
- Pass them in:
  - Add an extra parameter to `mw_get_field()`
  - Modify all the callers to pass in the correct function
  - Save them in `struct EnterWindowData`

- Refactor `op_editor_complete()` to use the passed function

- Tidy:
  - Find better homes for each function
  - Eliminate members from `struct EnterWindowData`
  - Eliminate parameters from `mw_get_field()`

**Reply by flatcap** [reply]
See:  #4025 editor: move completion functions
